pyiron/base/generic/hdfio.py

```diff
@@ -1271,31 +1308,27 @@ def to_object(self, object_type=None, **qwargs):
             qwargs["project"] = self.__class__(
                 project=self._create_project_from_hdf5(), file_name=qwargs["job_name"]
             )
-        if "TYPE" not in self.list_nodes():
-            if object_type is None:
-                raise ValueError(
-                    "Objects can be only recovered from hdf5 if TYPE is given"
-                )
-            else:
-                obj_type = object_type
-        else:
-            obj_type = self.get("TYPE")
-            if object_type is not None:
-                if object_type != obj_type:
-                    raise ValueError(
-                        "Object type in hdf5-file must be identical to input parameter"
-                    )
-        new_obj = self.create_object(obj_type, **qwargs)
-        if obj_type != str(type(new_obj)):  # Backwards compatibility
-            self["TYPE"] = str(type(new_obj))
-        if obj_type != "<class 'pyiron.atomistics.structure.atoms.Atoms'>":
-            new_obj.from_hdf()
-        else:
-            new_obj.from_hdf(
-                hdf=self.open(".."),
-                group_name=self.h5_path.split('/')[-1]
+        if "TYPE" not in self.list_nodes() and class_name is None:
+            raise ValueError(
+                "Objects can be only recovered from hdf5 if TYPE is given"
+            )
+        elif class_name is not None and class_name != self.get("TYPE"):
+            raise ValueError(
+                "Object type in hdf5-file must be identical to input parameter"
             )
-        return new_obj
+        class_name = class_name or self.get("TYPE")
+        class_object = self.import_class(class_name)
+
+        # TODO: what does that mean?
+        if class_name != str(class_object):  # Backwards compatibility
+            self["TYPE"] = str(class_object)
+
+        obj = self.make_from_hdf(class_object, **qwargs)
+        obj.from_hdf(
```

Comment:
This part does not work for the `Atoms()` class

Reply:
I tested it and it works when I call `Atoms.to_hdf` on a `ProjectHDFio`, then reload it and call `to_object`.  `make_from_hdf` takes care to pass `job_id` and `project` only when needed.